### ISSUE TITLE
StorageAnalyzer slows system responsiveness after opening heavily populated media folders

### ISSUE THREAD

**Issue by dgw** [d4rken-org/sdmaid-se#1166]
**Describe the bug**
System becomes slow or unresponsive after StorageAnalyzer has been used to view a folder containing many (thousands) media files.

**To Reproduce**
1. Place 5-10 thousand (or more) media files into a single folder, e.g. `/storage/emulated/0/Pictures/TestFolder`
   - Optionally add a `.nomedia` file to this folder — I am not sure yet if this influences the behavior
2. Scan device with StorageAnalyzer
3. Browse the StorageAnalyzer results; system responsiveness should still be normal
4. Enter the test folder containing thousands of media files
5. Scrolling may or may not work; SD Maid SE should now be slow to react to Back button presses or other input
6. When I have been able to get StorageAnalyzer to back out of the problematic folder, system responsiveness _has not_ recovered, as if some background task(s) spawned by opening the folder stay(s) running

**Expected behavior**
Regardless of how many files are in a given folder, browsing StorageAnalyzer results should remain responsive.

**Smartphone (please complete the following information):**
 - Device: ASUS ROG Phone 3 _and_ OnePlus Pad
 - OS: Android 12 _and_ OxygenOS 14 (yes, it's just skinned Android)
 - Root status: No
 - SD Maid 2/SE version: 0.24.0-beta0

**Debug log**
_I will be happy to add this later, if needed; it takes quite a while to scan my devices_

**Screenshots/Video**
_As above, happy to add if needed._

**Additional context**
Generally I end up restarting the whole device to get responsiveness back. I can usually do so through the normal menus, with patience, but occasionally it has been bad enough that I considered just holding the power button to force shutdown.

**Comment by d4rken** [comment]
If you disable previews in SD Maids settings, does it perform better?

**Comment by dgw** [comment]
With file previews disabled, StorageAnalyzer behaves much better in my test folder! The high file count on its own clearly isn't the problem.

Do you need me to check some other behavior? Maybe I could try removing the `.nomedia` file and let MediaServer index the folder overnight, then check StorageAnalyzer again with previews enabled? AFAIK there is no thumbnail cache for nomedia folders, which could be forcing thumbnail generation from scratch on all 5000 files. 

Improvement opportunity there would be only to help the UI stay responsive while thumbs are created, since it looks like an external library called `coil` deals with the image/video files themselves. I wasn't able to follow the Kotlin code deep enough to see if it takes advantage of the system thumbnail cache, but even loading thousands of pre-generated thumbs could be slow.

**Comment by d4rken** [comment]
The `.nomedia` file shouldn't make a difference, `coil` ignores that and in my implementation also does not use any system media cache (using that would cause unwanted cross-interactions if we are cleaning caches, it would be bad to use caches too).

> but even loading thousands of pre-generated thumbs could be slow.

This should not be a concern as it (should) only load thumbnails for files that are visible in the list (+/- a few extra at the start and at the end).

I tested this on my phone with 5k media files in the WhatsApp folder and it did only load those items that were visible. Fast scrolling even cancels loading for items no longer in view. 

Is this on an external sdcard and could it be possible that the sdcard is very slow and these operations hangup the whole OS because the I/O operations are piling up?

**Comment by dgw** [comment]
I haven't managed to reproduce the extreme system-wide GUI lag I saw before, so let me strikeout that part of the original issue until/unless I can make it happen again.

> Is this on an external sdcard and could it be possible that the sdcard is very slow and these operations hangup the whole OS because the I/O operations are piling up?

I thought of storage speed as a possible factor. Neither device I tested before opening the issue even supports expandable storage (😡 modern Android designs), but the unresponsiveness is more noticeable on the cheaper OnePlus device. Confoundingly, though, both advertise UFS 3.1 storage in their hardware specifications—so I'm inclined to think the difference is attributable to the different SoCs instead (Dimensity 9000 vs. Snapdragon 865).

> I tested this on my phone with 5k media files in the WhatsApp folder and it did only load those items that were visible. Fast scrolling even cancels loading for items no longer in view.

The faster device (ROG 3/SD865) is able to scroll around a 20k-file folder with no issues as you describe here, except at the beginning of the list where StorageAnalyzer sorts the several-hundred-MB video files my test folder contains.

If I fast-scroll away from the huge video files down to single-digit-MB images, as soon as I can after opening the huge folder, I can spot thumbs being generated in quick sequence after some waiting. The delay suggests that thumbs are generated sequentially; and that once a large file starts being thumbnailed, the operation perhaps can't be canceled?

Maybe my earlier tests actually made the issue worse by queueing additional large videos, since I was scrolling not very far down, only around the top of the list where all the biggest (and therefore likely most expensive) files are.

----

_While testing the scroll behavior, I found and replaced a couple of files that hadn't downloaded correctly and were sitting at 0 bytes. Normally I'd have no reason to look at the small end of the analyzer's report for a folder, so thanks for the excuse to play around!_

**Comment by d4rken** [comment]
> If I fast-scroll away from the huge video files down to single-digit-MB images, as soon as I can after opening the huge folder, I can spot thumbs being generated in quick sequence after some waiting. The delay suggests that thumbs are generated sequentially; and that once a large file starts being thumbnailed, the operation perhaps can't be canceled?

Sounds like thumbnailing the big videos is causing the slow down :thinking:

I thought they are being created in parallel, but I'll have to check, I think it's just using `coil`'s default setting at the moment. I just wrote a module so that coil can load the images through SD Maid's access methods (`LocalGateway`/`SAFGateway` etc).

**Comment by d4rken** [comment]
I've made further improvements across various updates, but there is no specific "bug" here except that loading video thumbnails is resource intensive. Bit out of ideas :thinking:

**Comment by dgw** [comment]
If there are no changes related to this after 1.2.6-rc0, I'm happy to retest in the next day or two. Otherwise I'll be on deck to retest after the next F-droid release, so hopefully we can close this out as "good enough" 😸

**Comment by d4rken** [comment]
Tried one more thing and played with the parallelism count. If I limit the default scheduler to anything below the maximum core count it no longer blocks the UI operation of going to the previous screen.

I'm closing this with #1416 should be in v1.2.7+.

**Comment by dgw** [comment]
In 1.2.7 I can confirm the UI now remains mostly responsive while thumbnails are being generated.

The speed of that process could be improved, and it seems maybe there is _excessive_ parallelism happening there based on how long it takes to show any thumbnail, but the UI doesn't block any more. It does seem to try reading _all_ files in the folder at once (but I don't know of an Android equivalent to Linux `lsof` or Windows' ResMon that would help me confirm this hypothesis).

If `coil` currently needs to read the entire file before returning a thumbnail, reducing that requirement to only the first _n_ seconds would be a possible next step to further improve perceived responsiveness in this view. My gallery app (Fossify) seems to have less trouble with video thumbnails in the same folder, but it also appears to generate them serially (which would be less taxing on the flash storage).

I won't create any follow-up issue for this unless invited. My original report was about _the UI freezing_, and that's been resolved; tweaking thumbnail generation would be an entirely new can of worms and let's not start that unless you want to!

**Comment by dgw** [comment]
> > The speed of that process could be improved, and it seems maybe there is _excessive_ parallelism happening there based on how long it takes to show any thumbnail, but the UI doesn't block any more.
> 
> We could further optimize this, but I did some testing and could not find any significant difference as long as it was MAX_CORES - 1 to stop the UI from becoming unresponsive.
> 
> https://github.com/d4rken-org/sdmaid-se/blob/bf4a0035fbf09ab43d8c08e564a12bc975473f0b/app/src/main/java/eu/darken/sdmse/common/coil/CoilModule.kt#L53-L57
> 
> I'm not sure if the CPU or the flash controller is the bottleneck here.

On the OnePlus Pad it could be both… There is a noticeable difference when comparing to ROG Phone 3, which has both a higher end CPU _and_ faster flash.

At some point I should simply take responsibility for choosing a cheap-ish tablet with MediaTek silicon… 😅

> > It does seem to try reading _all_ files in the folder at once (but I don't know of an Android equivalent to Linux `lsof` or Windows' ResMon that would help me confirm this hypothesis).
> 
> It's not all files in a folder, but all visible rows in the list + a few at the top and the bottom, i.e. any rows of the recyclerview being bound.

That's still several dozen files at the OPP's screen size, and the StorageAnalyzer tool starts right off with the biggest ones—meaning _plenty_ of flash load if `coil` does indeed need to load all of the video file for thumbnailing.

> Maybe Coil will improve upon video thumbnail decoding in the future? There is a major v3 being in the works.

I thought about asking whether v3 would be worth testing! One of the issues about video performance that I found over on their repo mentioned speed improvements coming in 3.0. Will keep an eye on your changelogs here, to hope for further speedup whenever `coil` gets updated.

Thanks for coming back to this issue, btw! I'm perfectly happy to leave it alone now; don't think there's much point in either of us chasing what's left any further until the `coil` v3 release. 🙂
